crates/hir-def/src/item_tree/lower.rs

```diff
@@ -605,6 +605,41 @@ impl<'a> Ctx<'a> {
         generics.fill(&self.body_ctx, node);
 
         generics.shrink_to_fit();
+
+        if let Some(params) = node.generic_param_list() {
+            let params_by_name: FxHashMap<_, _> = params
+                .generic_params()
+                .filter_map(|param| {
+                    let name = match &param {
+                        ast::GenericParam::ConstParam(param) => param.name()?.as_name(),
+                        ast::GenericParam::LifetimeParam(param) => {
+                            Name::new_lifetime(&param.lifetime()?)
+                        }
+                        ast::GenericParam::TypeParam(param) => param.name()?.as_name(),
+                    };
+                    Some((name, param))
+                })
+                .collect();
+            for (idx, param) in generics.type_or_consts.iter() {
+                if let Some(name) = param.name() {
+                    if let Some(param) = params_by_name.get(name) {
+                        self.add_attrs(
+                            idx.into(),
+                            RawAttrs::new(self.db.upcast(), param, self.hygiene()),
+                        );
+                    }
+                }
+            }
+            for (idx, param) in generics.lifetimes.iter() {
+                if let Some(param) = params_by_name.get(&param.name) {
+                    self.add_attrs(
+                        idx.into(),
+                        RawAttrs::new(self.db.upcast(), param, self.hygiene()),
+                    );
+                }
+            }
+        }
```

Comment:
I'm not sure of a better way to store the attributes associated with generic parameters. Generic parameters are stored in `generics` (yielding the `Idx<_>`s we need here) within `generics.fill()`, but `fill()` can't call `self.add_attrs()` because it doesn't have access to `self` (which is of a private type). Therefore, this creates a mapping from generic parameter name to the parameter as `&dyn HasAttrs` in order to call `add_attrs()` for each of the parameters within this function. This works, but unnecessarily allocates a `HashMap` and makes a second pass through all of the generic parameters.

Reply:
`generics.fill` is only being called here, so we might as well give it a callback parameter (or two rather) like `impl FnMut(Idx<..>, ast::TypeOrConstParam)` and `impl FnMut(Idx<..>, ast::LifetimeParam)`